src/node/stage/approved.rs

```diff
@@ -1539,7 +1541,10 @@ impl Approved {
             self.churn_in_progress = false;
             core.msg_filter.reset();
 
-            if new_is_elder {
+            // During the handling of OurKey, there is only section_key_index got updated, meanwhile
+            // the edlers info doesn't got updated. Calling `promote_and_demote_elders` will make
+            // un-necessary DKG process of same participants but different section_key_index.
+            if new_is_elder && (old_elders_info != new_elders_info) {
```

Comment:
Normally this would be handled by `SectionUpdateBarrier` which makes sure we only update our shared state if the new elders info is in sync with the section chain. That means that the current elders info must be signed by the key previous to the last key. This is the case always except for the very first elders info, which is signed by the last key. This is because there is no previous key in that case. There is a special handling in `SectionUpdateBarrier` for this but it assumes that there is only one elders in that case. This assumption does not hold in some of the unit tests, because we create multiple elders but still have only one key in the chain.

I think a better fix for this would be one of these:
1. Either modify the unit tests so the initial section chain has more than one key
2. Or move this logic to `SectionUpdateBarrier::take`: https://github.com/maidsafe/routing/blob/master/src/section/section_update_barrier.rs#L100

The reason why I think it would be better is that the purpose of `SectionUpdateBarrier` is specifically to only allow complete updates to the shared state. Having additional check for that here can cause confusion for people reading the code.

Reply:
> Normally this would be handled by SectionUpdateBarrier which makes sure we only update our shared state if the new elders info is in sync with the section chain

Reply:
>Normally this would be handled by SectionUpdateBarrier which makes sure we only update our shared state if the new elders info is in sync with the section chain

You mean the check at https://github.com/maidsafe/routing/blob/master/src/node/stage/approved.rs#L1466-1468 ?

However, it seems during the handling of sync, there is no such check? https://github.com/maidsafe/routing/blob/master/src/node/stage/approved.rs#L593

Reply:
That's right, good point. Maybe we should address that, but I don't think that's the issue here. If I'm understanding it correctly, the problem is we reach conssesus on `OurKey` while having only one key in the section chain yet. So the `SectionUpdateBarrier` checks and sees that the current elders info is signed with the previous key and considers the update as complete even though we are still expecting a consensus on `SectionInfo`.

Reply:
> the problem is we reach conssesus on OurKey while having only one key in the section chain yet. So the SectionUpdateBarrier checks and sees that the current elders info is signed with the previous key and considers the update as complete even though we are still expecting a consensus on SectionInfo

Ok. I will check about this.

> Maybe we should address that

right, I will have a check to see whether addressing that will get something fixed

Reply:
> right, I will have a check to see whether addressing that will get something fixed

I doubt it. Non-malicious nodes should never send a `Sync` with incomplete `SharedState`. So maybe all that's necessary is reject `Sync`s if the contained shared state is not complete (that is, it's elders info is not signed with the previous key).